haystack/nodes/preprocessor/splitter.py

```diff
@@ -0,0 +1,755 @@
+# pylint: disable=unnecessary-lambda-assignment
+from typing import List, Optional, Tuple, Union
+
+try:
+    from typing import Literal, get_args
+except ImportError:
+    from typing_extensions import Literal, get_args  # type: ignore
+
+import logging
+import re
+import warnings
+from functools import partial
+from pathlib import Path
+from copy import deepcopy
+from pickle import UnpicklingError
+
+import nltk
+from tqdm.auto import tqdm
+from nltk import NLTKWordTokenizer
+from nltk.tokenize.api import TokenizerI
+from transformers import PreTrainedTokenizer
+
+from haystack.nodes.base import BaseComponent
+from haystack.schema import Document
+from haystack.nodes.preprocessor.merger import DocumentMerger
+from haystack.modeling.model.feature_extraction import FeatureExtractor
+
+
+logger = logging.getLogger(__name__)
+
+
+REGEX_METACHARS = r".^$*+?{}[]\|()"
+
+
+SplitBy = Literal["character", "token", "word", "sentence", "paragraph", "page", "regex"]
+
+
+class DocumentSplitter(BaseComponent):
+    """
+    Splits documents into smaller, shorter documents.
+
+    Can split on different units ('character', 'token', 'word', 'sentence', 'paragraph', 'page', or 'regex'),
+    at different lengths, and include some overlap across the splits.
+
+    It can also properly assign page numbers and re-assign headlines found in the metadata of the parent document
+    to each split document.
+
+    No char is lost in splitting, not even whitespace, and all headlines are preserved, However, text and headlines
+    duplication may occur if `split_overlap>0`.
+    """
+
+    outgoing_edges = 1
+
+    def __init__(
+        self,
+        split_by: SplitBy,
+        split_length: int,
+        split_regex: Optional[str] = None,
+        split_overlap: int = 0,
+        max_chars: int = 2000,
```

Comment:
I'm sure there's a good motivation for making `max_chars` obligatory, but my initial inclination would be to make it optional, especially if `max_tokens` itself becomes a hard cap; likewise in `NewPreProcessor`. (Perhaps I'm missing something here?) Alternatively, if `max_chars` remains obligatory then I think I'd make the default value somewhat higher (maybe 5k? but just spitballing), to minimize the likelihood of someone accidentally splitting their documents more than they intended because they didn't realize they needed to specify `max_chars`. That being said, the current code gives a really nice and clear warning when this happens so maybe this isn't a big deal 🙂

Reply:
I think I agree with @robpasternak here in making `max_chars` optional in the same way as `max_tokens`. For example, I think there are uses cases where large documents are desirable such as BM25 sparse search for document retrieval.